db/schema_tables.cc

```diff
@@ -1053,12 +1053,12 @@ static void merge_tables_and_views(distributed<service::storage_proxy>& proxy,
             std::vector<future<>> notifications;
             notifications.reserve(tables_diff.size() + views_diff.size());
             auto notify = [&] (auto& r, auto&& f) { boost::range::transform(r, std::back_inserter(notifications), f); };
+            notify(tables_diff.dropped, [&] (auto&& dt) { return mm.notify_drop_column_family(dt.schema); });
```

Comment:
I think the order is still not enforced. 

notify() only collects futures, doesn't wait for them to complete and `notify_drop_column_family()` is an async operation that spawns a new seastar thread. So all notifications run in parallel and drop notification may still be reordered against the matching create notifications.

Reply:
If I launch several `seastar::thread` instances that otherwise do not have preemption points, can they be reordered in their execution? I thought that in seastar case it would be a plain queue if all threads have the same scheduling group. In that case it will still work fine, because the first and most important preemption point in `drop_view` is taking a semaphore inside `view_builder`, which ensures serialization of bookkeeping requests, so I'd expect all drops to be queued on that semaphore before all creations (and the semaphore itself ensures ordering as well, being a fifo).

But, if my assumptions are incorrect or use too much internal details, I can just split this `when_all` into two (or, for symmetry, three) `when_alls` and ensure ordering. Would you prefer it this way?

Reply:
Actually, I can't see any huge benefit in running these in a `when_all`, it doesn't look like a super hot performance-sensitive place, and things like `drop_view` end up serialized by the view builder anyway. Perhaps I'll just drop `when_all` and wait for the futures in order, one by one?

Reply:
Ok, nevermind my previous comment, I need to at least group all table drops, view drops, etc. in a `when_all`, otherwise the concurrency will likely be too small for a large number of tables.

Reply:
I know see that I omitted the fact that the previous implementation uses reverse iterators (`rbegin` and `rend`)... I'll rework this so that the order is ensured and retest.

Reply:
I think it's safer to put an explicit barrier by waiting for the drop listeners.

The abstract `migration_listener` interface specifies that listeners run in a thread, so presumably, some listeners could defer.

`notify_drop_view()` runs `on_drop_view()` listeners in a loop. If any of the listeners defers, later `on_drop_view()` may be deferred until after some `on_create_view()`.

Reply:
I'm resolving this conversation, because after the force push the code it applied to is no longer here. @tgrabiec can you take a look now? I explicitly set the order of when deletions, creations and changes are notified, with a synchronization point after each one, it also allowed me to get rid of the additional vector.

Verified via running `materialized_views_test.TestMaterializedViews.mv_populating_from_existing_data_during_node_stop_test` multiple times in `release` mode, both with and without additional debug logs to see what the order is.